lewis/adapters/modbus.py

```diff
@@ -0,0 +1,405 @@
+        :return: bool True if this frame is structurally valid.
+        """
+        conditions = [
+            self.protocol_id == 0,  # Modbus always uses protocol 0
+            2 <= self.length <= 260,  # Absolute length limits
+            len(self.data) == self.length - 2,  # Total length matches data length
+        ]
+        return all(conditions)
+
+    def create_exception(self, code):
+        """
+        Create an exception frame based on this frame.
+
+        :param code: Modbus exception code to use for this exception
+        :return: ModbusTCPFrame instance that represents an exception
+        """
+        frame = deepcopy(self)
+        frame.length = 3
+        frame.fcode += 0x80
+        frame.data = bytearray(chr(code))
+        return frame
+
+    def create_response(self, data=None):
+        """
+        Create a response frame based on this frame.
+
+        :param data: Data section of response as bytearray. If None, request data section is kept.
+        :return: ModbusTCPFrame instance that represents a response
+        """
+        frame = deepcopy(self)
+        if data is not None:
+            frame.data = data
+        frame.length = 2 + len(frame.data)
+        return frame
+
+
+class ModbusProtocol(object):
+    def __init__(self, sender, databank):
+        self._buffer = bytearray()
+        self._databank = databank
+        self._send = lambda req: sender(req.to_bytearray())
+
+        self._handlermap = {
+            0x01: self._handle_read_coils,
+            0x02: self._handle_read_discrete_inputs,
+            0x03: self._handle_read_holding_registers,
+            0x04: self._handle_read_input_registers,
+            0x05: self._handle_write_single_coil,
+            0x06: self._handle_write_single_register,
+            0x0F: self._handle_write_multiple_coils,
+            # 0x10: self._handle_write_multiple_registers,
+        }
+
+    def process(self, data=None):
+        """
+        Process as much of current buffer as possible.
+
+        :param data: Optionally append given data to buffer prior to parsing
+        """
+        self._buffer.extend(bytearray(data or ''))
+
+        for request in self._buffered_requests():
+            handler = self._get_handler(request.fcode)
+            response = handler(request)
+            self._send(response)
+
+    def _buffered_requests(self):
+        """Generator to yield all complete modbus requests in the buffer"""
+        try:
+            yield ModbusTCPFrame(self._buffer)
+        except EOFError:
+            pass
+
+    def _get_handler(self, fcode):
+        """Return handler with signature handler(request) for function code fcode"""
+        return self._handlermap.get(
+            fcode,
+            lambda req: req.create_exception(MBEX.ILLEGAL_FUNCTION)
+        )
+
+    def _handle_read_coils(self, request):
+        """
+        Handle request as per Modbus Application Protocol v1.1b3:
+        Section 6.1 - (0x01) Read Coils
```

Comment:
I think for now it's okay, there could be a link to the PDF or the/a page where the document can be obtained.